File: python/computer.py
from abc import ABC, abstractmethod
from input_device import InputDevice, Keyboard, Mouse
from storage_device import StorageDevice, SSD, HDD
from processing_device import ProcessingDevice, CPU, GPU
from output_device import OutputDevice, Monitor, Printer
from wifi_device import WiFiDevices
from bluetooth_device import BluetoothDevices
from brands import Dell, IBM, Lenovo, HP, Brand


# Define Computer class
class Computer(ABC):
    def __init__(
        self,
        input_device: InputDevice,
        storage_device: StorageDevice,
        processing_device: ProcessingDevice,
        output_device: OutputDevice,
        brand: Brand,
    ):
        self._input_device = input_device
        self._storage_device = storage_device
        self._processing_device = processing_device
        self._output_device = output_device
        self.brand = brand

    @abstractmethod
    def manufacture(self) -> str:
        pass

    def boot(self):
        self.brand.boot()
    
    def set_input(self, device: InputDevice) -> None:
        self._input_device = device

    def set_storage(self, device: StorageDevice) -> None:
        self._storage_device = device

    def set_processing(self, device: ProcessingDevice) -> None:
        self._processing_device = device

    def set_output(self, device: OutputDevice) -> None:
        self._output_device = device

    def input(self) -> str:
        return self._input_device.input()

    def storage(self) -> str:
        return f"{self._storage_device.store()} and {self._storage_device.retrieve()}"

    def processing(self) -> str:
        return self._processing_device.process()

    def output(self) -> str:
        return self._output_device.output()


# Refinned Abstractions


class DesktopComputer(Computer):
    def manufacture(self):
        return f"Desktop Manufactured by the  {self.brand.set_brand()}"


class WalltopComputer(Computer):
    def manufacture(self):
        return f"Walltop Manufactured by the  {self.brand.set_brand()}"


class TabletComputer(Computer):
    def manufacture(self):
        return f"Tablet Manufactured by the  {self.brand.set_brand()}"


class PalmtopComputer(Computer):
    def manufacture(self):
        return f"Palmtop Manufactured by the  {self.brand.set_brand()}"


# Laptop extends Computer and implements wifi and Bluetooth interfaces
class Laptop(Computer, WiFiDevices, BluetoothDevices):

    def manufacture(self):
        return f"Laptop Manufactured by the  {self.brand.set_brand()}"

    def connectToBluetooth(self) -> str:
        return "Bluetooth connected"

    def disconnectFromBluetooth(self) -> str:
        return "Bluetooth disconnected"

    def connectToWifi(self) -> str:
        return "WiFi connected"

    def disconnectFromWifi(self) -> str:
        return "WiFi disconnected"


# Example usage
keyboard = Keyboard()
mouse = Mouse()
ssd = SSD()
cpu = CPU()
monitor = Monitor()


# computer = Computer(keyboard, ssd, cpu, monitor)
# print("=================Computer===============")
# print(computer.input())  # Outputs: Keyboard input
# print(computer.storage())  # Outputs: SSD storing data and SSD retrieving data
# print(computer.processing())  # Outputs: CPU processing data
# print(computer.output())  # Outputs: Monitor output
# # Change the input device
# computer.set_input(mouse)
# print(computer.input())  # Outputs: Mouse input
# # Create a computer without WiFi and Bluetooth support


Laptop1 = Laptop(keyboard, ssd, cpu, monitor, Dell())
Laptop2 = Laptop(keyboard, ssd, cpu, monitor, Lenovo())
Laptop3 = Laptop(keyboard, ssd, cpu, monitor, HP())
Laptop4 = Laptop(keyboard, ssd, cpu, monitor, IBM())
Laptop4.boot()

walltop1 = WalltopComputer(keyboard, ssd, cpu, monitor, Lenovo())
walltop2 = WalltopComputer(keyboard, ssd, cpu, monitor, Dell())
walltop3 = WalltopComputer(keyboard, ssd, cpu, monitor, IBM())
walltop4 = WalltopComputer(keyboard, ssd, cpu, monitor, HP())
walltop4.boot()

palmtop1 = PalmtopComputer(keyboard, ssd, cpu, monitor, IBM())
palmtop2 = PalmtopComputer(keyboard, ssd, cpu, monitor, Dell())
palmtop3 = PalmtopComputer(keyboard, ssd, cpu, monitor, HP())
palmtop4 = PalmtopComputer(keyboard, ssd, cpu, monitor, Lenovo())
palmtop4.boot()

desktop1 = DesktopComputer(keyboard, ssd, cpu, monitor, HP())
desktop2 = DesktopComputer(keyboard, ssd, cpu, monitor, Dell())
desktop3 = DesktopComputer(keyboard, ssd, cpu, monitor, IBM())
desktop4 = DesktopComputer(keyboard, ssd, cpu, monitor, Lenovo())
desktop2.boot()

print("=================Laptop===============")
print(Laptop1.manufacture())
print(Laptop2.manufacture())
print(Laptop3.manufacture())
print(Laptop4.manufacture())
print("")

print("=================Walltop===============")
print(walltop1.manufacture())
print(walltop2.manufacture())
print(walltop3.manufacture())
print(walltop4.manufacture())
print("")

print("=================Palmtop===============")
print(palmtop1.manufacture())
print(palmtop2.manufacture())
print(palmtop3.manufacture())
print(palmtop4.manufacture())
print("")

print("=================Desktop===============")
print(desktop1.manufacture())
print(desktop2.manufacture())
print(desktop3.manufacture())
print(desktop4.manufacture())
print("")

File: typescript/computer.ts
""
import InputDevice, {Mouse, Keyboard} from "./inputDevices";
import StorageDevice, {SSD, HDD} from "./storageDevices";
import ProcessingDevice, {CPU, GPU} from "./processingDevices";
import OutputDevice, {Monitor, Printer} from "./outputDevices";
import Brand, {IBM,Dell,Hp,Lenovo} from "./brands";


// WiFi & Bluetooth Devices

interface WiFi {
    connectToWifi();
    disconnectFromWiFi();
}

interface Bluetooth {
    connectToBluetooth()
    disconnectFromBluetooth()
}


// The Computer Interface
abstract class Computer {
    private inputDevice: InputDevice;
    private storageDevice: StorageDevice;
    private processingDevice: ProcessingDevice;
    private outputDevice: OutputDevice;
    public brand: Brand;

    constructor(
        inputDevice: InputDevice,
        storageDevice: StorageDevice,
        processingDevice: ProcessingDevice,
        outputDevice: OutputDevice,
        brand: Brand,
        wifi?: WiFi,
        bluetooth?: Bluetooth
    ) {
        this.inputDevice = inputDevice;
        this.storageDevice = storageDevice;
        this.processingDevice = processingDevice;
        this.outputDevice = outputDevice;
        this.brand = brand;
    }

    abstract manufacturer(): string 

    boot(){
        this.brand.boot();
    }

    setInputDevice(device: InputDevice): void {
        this.inputDevice = device;
    }

    setStorageDevice(device: StorageDevice): void {
        this.storageDevice = device;
    }

    setProcessingDevice(device: ProcessingDevice): void {
        this.processingDevice = device;
    }

    setOutputDevice(device: OutputDevice): void {
        this.outputDevice = device;
    }



    Input(): string {
        return this.inputDevice.input();
    }
    
    Storage(): string {
        return this.storageDevice.store() + " and " + this.storageDevice.retrieve();
    }

    Processing(): string {
        return this.processingDevice.process();
    }

    Output(): string {
        return this.outputDevice.output();
    }

}

class Desktop extends Computer {
    manufacturer(): string {
        return "Desktop maunfactured by: " + this.brand.setBrand();
    }
}

class Laptop extends Computer implements WiFi, Bluetooth{

    manufacturer(): string {
        return "Laptop maunfactured by: " + this.brand.setBrand();
    }

    connectToWifi(): string {
        return "WiFi connected";
    }
    disconnectFromWiFi(): string {
        return "WiFi disconnected";
    }

    connectToBluetooth(): string {
        return "Bluetooth connected";
    }
    disconnectFromBluetooth(): string {
        return "Bluetooth disconnected";
    }

}



// Example usage
let keyboard = new Keyboard();
let mouse = new Mouse();
let ssd = new SSD();
let hdd = new HDD();
let cpu = new CPU();
let gpu = new GPU();
let monitor = new Monitor();
let printer = new Printer();
let hp = new Hp();
let lenovo = new Lenovo();



// Desktop does not support WiFi and Bluetooth technologies
// let computer = new Computer(keyboard, ssd, cpu, monitor);
// console.log("========= COMPUTER =================================");
// console.log(computer.Input());      // Outputs: Keyboard input
// computer.setInputDevice(mouse);
// console.log(computer.Input());      // Outputs: Mouse input

// computer.setStorageDevice(hdd);
// console.log(computer.Storage());      // Outputs: HDD input

// console.log(computer.Processing()); // Outputs: CPU processing data
// computer.setProcessingDevice(gpu);
// console.log(computer.Processing()); // Outputs: GPU processing data

// console.log(computer.Output());    // Outputs: Monitor output
// computer.setOutputDevice(printer);
// console.log(computer.Output());




// Laptop supports WiFi and Bluetooth technologies
let laptop = new Laptop(keyboard, ssd, cpu, monitor,hp);
console.log("========= LAPTOP =================================");

laptop.boot()

laptop.setInputDevice(keyboard);
console.log(laptop.Input());      // Outputs: Keyboard input
laptop.setInputDevice(mouse);
console.log(laptop.Input());      // Outputs: Mouse input

laptop.setStorageDevice(ssd);
console.log(laptop.Storage());laptop.setStorageDevice(hdd);
console.log(laptop.Storage());


laptop.setProcessingDevice(cpu);
console.log(laptop.Processing());
laptop.setProcessingDevice(gpu);
console.log(laptop.Processing());

laptop.setOutputDevice(monitor);
console.log(laptop.Output());
laptop.setOutputDevice(printer);
console.log(laptop.Output());

console.log(laptop.connectToWifi());       // Outputs: WiFi connected
console.log(laptop.connectToBluetooth());  // Outputs: Bluetooth connected
console.log(laptop.manufacturer());  

let desktop = new Desktop(keyboard, ssd, cpu, monitor,lenovo);
console.log(desktop.manufacturer()); 